src/models/neural_networks/rnn.py

```diff
@@ -0,0 +1,175 @@
+                 batch_size: int = 1,
+                 pred_months: Optional[List[int]] = None,
+                 include_pred_month: bool = True) -> None:
+        super().__init__(data_folder, batch_size, pred_months, include_pred_month)
+
+        # to initialize and save the model
+        self.hidden_size = hidden_size
+        self.rnn_dropout = rnn_dropout
+        self.dense_features = dense_features
+        self.features_per_month: Optional[int] = None
+
+    def save_model(self):
+
+        assert self.model is not None, 'Model must be trained before it can be saved!'
+
+        model_dict = {
+            'state_dict': self.model.state_dict(),
+            'features_per_month': self.features_per_month,
+            'hidden_size': self.hidden_size,
+            'rnn_dropout': self.rnn_dropout,
+            'dense_features': self.dense_features,
+            'include_pred_month': self.include_pred_month
+        }
+
+        torch.save(model_dict, self.model_dir / 'model.pkl')
+
+    def _initialize_model(self, x_ref: Tuple[torch.Tensor, ...]) -> nn.Module:
+        self.features_per_month = x_ref[0].shape[-1]
+        return RNN(features_per_month=self.features_per_month,
+                   dense_features=self.dense_features,
+                   hidden_size=self.hidden_size,
+                   rnn_dropout=self.rnn_dropout,
+                   include_pred_month=self.include_pred_month)
+
+
+class RNN(nn.Module):
+    def __init__(self, features_per_month, dense_features, hidden_size,
+                 rnn_dropout, include_pred_month):
+        super().__init__()
+
+        self.include_pred_month = include_pred_month
+
+        self.dropout = nn.Dropout(rnn_dropout)
+        self.rnn = UnrolledRNN(input_size=features_per_month,
+                               hidden_size=hidden_size,
+                               batch_first=True)
+        self.hidden_size = hidden_size
+
+        dense_input_size = hidden_size
+        if include_pred_month:
+            dense_input_size += 12
+        dense_features.insert(0, dense_input_size)
+        if dense_features[-1] != 1:
+            dense_features.append(1)
+
+        self.dense_layers = nn.ModuleList([
+            nn.Linear(in_features=dense_features[i - 1],
+                      out_features=dense_features[i])
+            for i in range(1, len(dense_features))
+        ])
+
+        self.initialize_weights()
+
+    def initialize_weights(self):
+
+        sqrt_k = math.sqrt(1 / self.hidden_size)
+        for parameters in self.rnn.parameters():
+            for pam in parameters:
+                nn.init.uniform_(pam.data, -sqrt_k, sqrt_k)
+
+        for dense_layer in self.dense_layers:
+            nn.init.kaiming_uniform_(dense_layer.weight.data)
+            nn.init.constant_(dense_layer.bias.data, 0)
+
+    def forward(self, x, pred_month=None):
+
+        sequence_length = x.shape[1]
+
+        hidden_state = torch.zeros(1, x.shape[0], self.hidden_size)
```

Comment:
What's the difference between the `hidden state` and `cell_State`?

![Screenshot 2019-07-16 at 10 04 46](https://user-images.githubusercontent.com/21049064/61281224-33b41280-a7b1-11e9-82c7-7dd15393c093.png)


Is it the difference between `A` (the layers) and `C` the conveyor belt passing through the layers?

Reply:
Note the two arrows coming into the cell from the previous cell - one represents the cell state, the other the hidden state.

The cell state is intended to represent the "long term" information being captured. The hidden state is the output of the specific cell - `h_t`. Note that this is both the output of the cell, and is also passed to the next cell.